test/DaemonExec.test.ts

```diff
@@ -0,0 +1,167 @@
+import * as Docker from 'dockerode'
+
+import { promises as fs, createReadStream, createWriteStream } from 'fs'
+
+import { resolve, join } from 'path'
+
+import { tmpdir } from 'os'
+import {
+	VolumeDefinition,
+	listImages,
+	listContainers,
+	ensureImageImport,
+	createContainer,
+	attachStreams,
+	stopContainer,
+	removeContainer,
+} from '../src/Daemon/DaemonExec'
+
+interface ExpectedStatus {
+	Error: unknown
+	StatusCode: number
+}
+
+const docker = new Docker()
+const testImage = 'ubuntu:18.04'
+const bogusImage = 'lskjflskjflsdkfjslkfjsdlfsdjflksdj'
+const inputFile = 'package.json'
+
+describe('DaemonExec', () => {
+	let container: Docker.Container
+	let tempDir: string
+	let outputFile: string
+	let errorFile: string
+
+	it('pulls a new image', async () => {
+		const initialInfos = await listImages(docker)
+		if (
+			initialInfos.some((info) =>
+				info.RepoTags.some((tag) => tag === testImage),
+			)
+		) {
+			// if image is already imported, remove it
+			const image = docker.getImage(testImage)
+			await image.remove()
+		}
+
+		await ensureImageImport(docker, testImage)
+		const newInfos = await listImages(docker)
+
+		expect(
+			newInfos.some((info) => info.RepoTags.some((tag) => tag === testImage)),
+		).toBeTruthy()
+	}, 60000)
+
+	it('pulls an existing image', async () => {
```

Comment:
How do you know this actually quits early with the existing image as opposed to downloading it? Maybe Travis has a fast internet connection ;)

Add a comment to document your assumption that a new image can't get downloaded in less than 5 seconds.

---

Btw, don't actually spend time on this since we need to focus on the SSH daemon, but it's a good idea to add a unit test for this method (passing it a fake Docker) in addition to this integration test, so you can test that it's following the right logic. Usually unit tests are written first.

Reply:
This test is for the behavior when a requested image already exists. It is not necessarily testing whether it will download the image again, but whether the image exists after the method is called, and the method does not throw an error. It is a convenient coincidence that the image does not get downloaded.

Reply:
I see, this makes sense then.